%\VignetteEngine{knitr::knitr}
%\VignetteIndexEntry{R/qtlcharts User Guide}

# R/qtlcharts User Guide

[R/qtlcharts](http://kbroman.github.io/qtlcharts) is an
[R](http://www.r-project.org) package to create interactive charts for
QTL data, for use with [R/qtl](http://www.rqtl.org). The charts are
saved to a temporary file and opened in a web browser, though you may also
save them to a specific file, to be viewed later.

We are focusing development on the
[Chrome](http://www.google.com/chrome/browser) and
[Safari](https://www.apple.com/safari/) browsers. The graphs may also
be viewed in other browsers, such [Opera](http://www.opera.com/) and
[Firefox](http://www.mozilla.org/en-US/firefox/new/), but it can be
hard to accommodate all possible browser differences.

Information on installing R/qtlcharts is available
[here](http://kbroman.github.io/qtlcharts/pages/installation.html); it
requires installation of the packages [R/qtl](http://www.rqtl.org),
[jsonlite](http://cran.r-project.org/web/packages/jsonlite/), and
[devtools](https://github.com/hadley/devtools).

## Basic use

You first need to load the package.

```{r load_library}
library(qtlcharts)
```

Now let's begin by considering the function `iplotCorr`, which creates
a heatmap of a correlation matrix, linked to scatterplots of the
underlying variables.

We'll first load the `geneExpr` dataset included with the R/qtlcharts package.

```{r geneExpr_data}
data(geneExpr)
```

This is a list with two components. The first, `geneExpr$expr`, is a
`r nrow(geneExpr$expr)` &times; `r ncol(geneExpr$expr)` matrix of gene
expression data. We can use `iplotCorr` with these data, as follows:

```{r iplotCorr_reordered, eval=FALSE}
iplotCorr(geneExpr$expr, reorder=TRUE)
```

This will open ...

![iplotCorr example](Figs/iplotCorr1.png)


Scatterplot colored by genotype, contained in the vector `geneExpr$genotype`.

```{r iplotCorr_reordered_with_genotype, eval=FALSE}
iplotCorr(geneExpr$expr, geneExpr$genotype, reorder=TRUE)
```

![iplotCorr example with scatterplot colored by genotype](Figs/iplotCorr2.png)


## Saving to a file

The interactive graphs produced by R/qtlcharts are, by default, saved
to a temporary file and then opened in the default web browser. If you
want to save a chart to a _particular_ file, just use the `file`
argument, as follows:

```{r iplotCorr_specify_file, eval=FALSE}
iplotCorr(geneExpr$expr, reorder=TRUE, file="~/Desktop/iplotCorr_example.html")
```

The chart will still be opened in the browser; to create the file
without opening it, use the argument `openfile=FALSE`, as follows:

```{r iplotCorr_specify_file_dont_open, eval=FALSE}
iplotCorr(geneExpr$expr, reorder=TRUE, file="~/Desktop/iplotCorr_example.html",
          openfile=FALSE)
```

Charts saved in this way will contain links to resources within the
R/qtlcharts package, and so will be viewable on the same system but
will not be portable to another system that doesn't have R/qtlcharts
installed, or that has the package installed in a different location.

To make the interactive graph portable to other computers (for
example, for sharing with collaborators), use the argument
`onefile=TRUE`, as follows:

```{r iplotCorr_make_portable, eval=FALSE}
iplotCorr(geneExpr$expr, reorder=TRUE, file="~/Desktop/iplotCorr_example.html",
          onefile=TRUE)
```

The file will be somewhat larger, but it will be self-contained and so
may be transferred and viewed on any computer.


## Including a chart in an R Markdown document

The interactive charts may be included within an
[R Markdown](https://www.rstudio.com/ide/docs/r_markdown)-based
report. See the separate `Rmarkdown` vignette:

```{r Rmarkdown_vignette, eval=FALSE}
vignette("Rmarkdown", "qtlcharts")
```




## Examples of the use of other charts

### `iplotMap`

`iplotMap` creates a (slightly) interactive genetic map. Hover over a
marker position to view the marker name.

```{r iplotMap_example, eval=FALSE}
data(hyper)
map <- pull.map(hyper)

iplotMap(map, title="iplotMap example")
```

### `iplotScanone`

`iplotScanone` creates an interactive chart with LOD
curves linked to estimated QTL effects.


```{r iplotScanone_example, eval=FALSE}
data(hyper)
hyper <- calc.genoprob(hyper, step=1)
out <- scanone(hyper)

iplotScanone(out, hyper, title="iplotScanone example")
```

### `iplotMScanone`

`iplotMScanone` creates a heatmap of LOD curves for a set of genome
scans (for each, for each of a time course of phenotypes), linked to a
plot of the individual genome scans, and also to the QTL effects.

```{r iplotMScanone_example, eval=FALSE}
data(grav)
grav <- calc.genoprob(grav, step=1)
grav <- reduce2grid(grav)
phecol <- seq(1, nphe(grav), by=5)
out <- scanone(grav, phe=phecol)
eff <- estQTLeffects(grav, phe=phecol, what="effects")

iplotMScanone(out, effects=eff,
              title="iplotMScanone example",
              chartOpts=list(eff_ylab="QTL effect"))
```


### `iplotCorr`

`iplotCorr` creates an heat map of a correlation matrix, linked to the
underlying scatterplots.

```{r iplotCorr_example, eval=FALSE}
data(geneExpr)

iplotCorr(geneExpr$expr, geneExpr$genotype,
          title = "iplotCorr example",
          chartOpts=list(cortitle="Correlation matrix",
                         scattitle="Scatterplot"))
```


### `iplotCurves`

`iplotCurves` creates a plot of a set of curves linked to one or two
scatterplots.

```{r iplotCurves_example, eval=FALSE}
data(grav)
times <- attr(grav, "time")
phe <- grav$pheno

iplotCurves(phe, times, phe[,times==2 | times==4], phe[,times==4 | times==6],
            title="iplotCurves example",
            chartOpts=list(curves_xlab="Time (hours)", curves_ylab="Root tip angle (degrees)",
                           scat1_xlab="Angle at 2 hrs", scat1_ylab="Angle at 4 hrs",
                           scat2_xlab="Angle at 4 hrs", scat2_ylab="Angle at 6 hrs"))
```

### `iboxplot`

`iboxplot` creates an interactive graph for a large set of box plots
(rendered as lines connecting the quantiles), linked to underlying
histograms.

```{r iboxplot_example, eval=FALSE}
# simulated data
n.ind <- 500
n.gene <- 10000
expr <- matrix(rnorm(n.ind * n.gene, (1:n.ind)/n.ind*3), ncol=n.gene)
dimnames(expr) <- list(paste0("ind", 1:n.ind),
                       paste0("gene", 1:n.gene))

iboxplot(expr, title="iboxplot example",
         chartOpts=list(xlab="Mice", ylab="Gene expression"))
```

### `iheatmap`

`iheatmap` creates an interactive heatmap of a numeric matrix, with
pixels linked to plots of horizontal and vertical slices.

```{r iheatmap_example, eval=FALSE}
# data to display
n <- 101
x <- y <- seq(-2, 2, len=n)
z <- matrix(ncol=n, nrow=n)
for(i in seq(along=x))
  for(j in seq(along=y))
    z[i,j] <- x[i]*y[j]*exp(-x[i]^2 - y[j]^2)

iheatmap(z, x, y, title = "iheatmap example")
```




## Full list of `chartOpts` options

There are a number of ways in which one may wish to customize the
interactive graphs produced by
[R/qtlcharts](http://kbroman.github.io/qtlcharts), such as different
colors or axis labels. Our main approach to allow such customization
is through a function argument, `chartOpts`. We touched on this
briefly above. The following is a complete listing of the
customization options for each chart.

### `iboxplot`

`width = 1000`: width of image in pixels

`height = 450`: height of image in pixels

`margin = list(left=60, top=20, right=60, bottom=40)`: margins in pixels (left, top, right, bottom)

`ylab = "Response"`: y-axis label

`xlab = "Individuals"`: x-axis label

`rectcolor = "#E6E6E6"`: color of background rectangle

`qucolors = NULL`: vector of colors for the quantile curves

`histcolors = c("blue", "red", "green", "MediumVioletRed", "black")`: vector of colors for selected histograms

### `iheatmap`

`htop = 500`: height of top charts in pixels

`hbot = 500`: height of bottom chart in pixels

`wleft = 500`: width of left charts in pixels

`wright = 500`: width of right chart in pixels

`margin = list(left=60, top=40, right=40, bottom=40, inner=5)`: margins in pixels (left, top, right, bottom, inner)

`axispos = list(xtitle=25, ytitle=30, xlabel=5, ylabel=5)`: position of axis labels in pixels (xtitle, ytitle, xlabel, ylabel)

`titlepos = 20`: position of chart title in pixels

`rectcolor = "#E6E6E6"`: color of background rectangle

`strokecolor = "slateblue"`: line color

`strokewidth = 2`: line width

`xlim = range(x)`: x-axis limits

`ylim = range(y)`: y-axis limits

`nxticks = 5`: no. ticks on x-axis

`xticks = NULL`: vector of tick positions on x-axis

`nyticks = 5`: no. ticks on y-axis

`yticks = NULL`: vector of tick positions on y-axis

`nzticks = 5`: no. ticks on z-axis

`zticks = NULL`: vector of tick positions on z-axis

`title = ""`: title for chart

`xlab = "X"`: x-axis label

`ylab = "Y"`: y-axis label

`zlab = "Z"`: z-axis label

`zthresh = NULL`: lower threshold for z-axis for plotting in heat map

`zlim = c(-max(z), 0, max(z))`: z-axis limits

`colors = c("slateblue", "white", "crimson")`: heat map colors (same length as `zlim`)

### `iplotCorr`

`height = 450`: height of each panel in pixels

`width = height`: width of each panel in pixels

`margin = list(left=70, top=40, right=5, bottom=70, inner=5)`: margins in pixels (left, top, right, bottom, inner)

`corcolors = c("darkslateblue", "white", "crimson")`: heat map colors (same length as `zlim`)

`zlim = c(-1, 0, 1)`: z-axis limits

`rectcolor = "#E6E6E6"`: color of background rectangle

`cortitle = ""`: title for heatmap panel

`scattitle = ""`: title for scatterplot panel

`scatcolors = NULL`: vector of point colors for scatterplot

### `iplotCurves`

`htop = 500`: height of curves chart in pixels

`hbot = 500`: height of scatterplots in pixels

`width = 1000`: width of chart in pixels

`margin = list(left=60, top=40, right=40, bottom=40, inner=5)`: margins in pixels (left, top, right, bottom, inner)

`axispos = list(xtitle=25, ytitle=30, xlabel=5, ylabel=5)`: position of axis labels in pixels (xtitle, ytitle, xlabel, ylabel)

`titlepos = 20`: position of chart title in pixels

`rectcolor = "#E6E6E6"`: color of background rectangle

`pointcolor = NULL`: vector of colors for points in scatterplots

`pointstroke = "black"`: color of line outline for points in scatterplots

`pointsize = 3`: size of points in scatterplots

`pointcolorhilit = NULL`: vector of colors for points in scatterplots, when highlighted

`pointsizehilit = 6`: zie of points in scatterplot, when highlighted

`strokecolor = NULL`: vector of colors of curves

`strokecolorhilit = NULL`: vector of colors of curves, when highlighted

`strokewidth = 2`: line width of curves

`strokewidthhilit = 2`: line widths of curves, when highlighted

`curves_xlim = NULL`: x-axis limits in curve plot

`curves_ylim = NULL`: y-axis limits in curve plot

`curves_nxticks = 5`: no. ticks on x-axis in curve plot

`curves_xticks = NULL`: vector of tick positions on x-axis in curve plot

`curves_nyticks = 5`: no. ticks on y-axis in curve plot

`curves_yticks = NULL`: vector of tick positions on y-axis in curve plot

`curves_title = ""`: title for curve plot

`curves_xlab = "X"`: x-axis label for curve plot

`curves_ylab = "Y"`: y-axis label for curve plot

`scat1_xlim = NULL`: x-axis limits in first scatterplot

`scat1_ylim = NULL`: y-axis limits in first scatterplot

`scat1_xNA = list(handle=true, force=false, width=15, gap=10)`: treatment of missing values for x variable in first scatterplot (handle=T/F, force=T/F, width, gap)

`scat1_yNA = list(handle=true, force=false, width=15, gap=10)`: treatment of missing values for x variable in first scatterplot (handle=T/F, force=T/F, width, gap)

`scat1_nxticks = 5`: no. ticks on x-axis in first scatterplot

`scat1_xticks = NULL`: vector of tick positions on x-axis in first scatterplot

`scat1_nyticks = 5`: no. ticks on y-axis in first scatterplot

`scat1_yticks = NULL`: vector of tick positions on y-axis in first scatterplot

`scat1_title = ""`: title for first scatterplot

`scat1_xlab = "X"`: x-axis label for first scatterplot

`scat1_ylab = "Y"`: y-axis label for first scatterplot

`scat2_xlim = NULL`: x-axis limits in second scatterplot

`scat2_ylim = NULL`: y-axis limits in second scatterplot

`scat2_xNA = list(handle=true, force=false, width=15, gap=10)`: treatment of missing values for x variable in second scatterplot (handle=T/F, force=T/F, width, gap)

`scat2_yNA = list(handle=true, force=false, width=15, gap=10)`: treatment of missing values for x variable in second scatterplot (handle=T/F, force=T/F, width, gap)

`scat2_nxticks = 5`: no. ticks on x-axis in second scatterplot

`scat2_xticks = NULL`: vector of tick positions on x-axis in second scatterplot

`scat2_nyticks = 5`: no. ticks on y-axis in second scatterplot

`scat2_yticks = NULL`: vector of tick positions on y-axis in second scatterplot

`scat2_title = ""`: title for second scatterplot

`scat2_xlab = "X"`: x-axis label for second scatterplot

`scat2_ylab = "Y"`: y-axis label for second scatterplot

### `iplotMap`

`width = 1000`: width of chart in pixels

`height = 600`: height of chart in pixels

`margin = list(left=60, top=40, right=40, bottom=40, inner=10)`: margins in pixels (left, top, right, bottom, inner)

`axispos = list(xtitle=25, ytitle=30, xlabel=5, ylabel=5)`: position of axis labels in pixels (xtitle, ytitle, xlabel, ylabel)

`titlepos = 20`: position of chart title in pixels

`ylim = NULL`: y-axis limits

`nyticks = 5`: no. ticks on y-axis

`yticks = NULL`: vector of tick positions on y-axis

`tickwidth = 10`: width of tick marks at markers, in pixels

`rectcolor = "#E6E6E6"`: color of background rectangle

`linecolor = "slateblue"`: color of lines

`linecolorhilit = "Orchid"`: color of lines, when highlighted

`linewidth = 3`: width of lines

`title = ""`: title for chart

`xlab = "Chromosome"`: x-axis label

`ylab = "Position (cM)"`: y-axis label

### `iplotScanone` (no effects)

`height = 450`: height of image in pixels

`width = 900`: width of image in pixels

`margin = list(left=60, top=40, right=40, bottom=40, inner=5)`: margins in pixels (left, top, right, bottom, inner)

`axispos = list(xtitle=25, ytitle=30, xlabel=5, ylabel=5)`: position of axis labels in pixels (xtitle, ytitle, xlabel, ylabel)

`titlepos = 20`: position of chart title in pixels

`ylim = NULL`: y-axis limits

`nyticks = 5`: number of ticks in y-axis

`yticks = NULL`: vector of tick positions for y-axis

`chrGap = 8`: gap between chromosomes in pixels

`darkrect = "#C8C8C8"`: color of darker background rectangle

`lightrect = "#E6E6E6"`: color of lighter background rectangle

`linecolor = "darkslateblue"`: line color for LOD curves

`linewidth = 2`: line width for LOD curves

`pointcolor = "#E9CFEC"`: color for points at markers

`pointsize = 0`: size of points at markers (default = 0 corresponding to no visible points at markers)

`pointstroke = "black"`: color of outer circle for points at markers

`title = ""`: title of chart

`xlab = "Chromosome"`: x-axis label

`ylab = "LOD score"`: y-axis label

`rotate_ylab = NULL`: indicates whether to rotate the y-axis label 90 degrees

### `iplotScanone` (`pxgtype="ci"`)

`height = 450`: height of image in pixels

`wleft = 700`: width of left panel in pixels

`wright = 300`: width of right panel in pixels

`margin = list(left=60, top=40, right=40, bottom=40, inner=5)`: margins in pixels (left, top, right, bottom, inner)

`lod_axispos = list(xtitle=25, ytitle=30, xlabel=5, ylabel=5)`: position of axis labels in pixels (xtitle, ytitle, xlabel, ylabel) in LOD curve panel

`lod_titlepos = 20`: position of title for LOD curve panel, in pixels

`chrGap = 8`: gap between chromosomes

`darkrect = "#C8C8C8"`: color of darker background rectangle

`lightrect = "#E6E6E6"`: color of lighter background rectangle

`lod_ylim = NULL`: y-axis limits in LOD curve panel

`lod_nyticks = 5`: number of ticks in y-axis in LOD curve panel

`lod_yticks = NULL`: vector of tick positions for y-axis in LOD curve panel

`lod_linecolor = "darkslateblue"`: line color for LOD curves

`lod_linewidth = 2`: line width for LOD curves

`lod_pointcolor = "#E9CFEC"`: color for points at markers in LOD curve panel

`lod_pointsize = 0`: size of points at markers (default = 0 corresponding to no visible points at markers)

`lod_pointstroke = "black"`: color of outer circle for points at markers in LOD curve panel

`lod_title = ""`: title of LOD curve panel

`lod_xlab = "Chromosome"`: x-axis label for LOD curve panel

`lod_ylab = "LOD score"`: y-axis label for LOD curve panel

`lod_rotate_ylab = NULL`: indicates whether to rotate the y-axis label 90 degrees, in LOD curve panel

`eff_ylim = NULL`: y-axis limits in effect plot panel

`eff_nyticks = 5`: number of ticks in y-axis in effect plot panel

`eff_yticks = NULL`: vector of tick positions for y-axis in effect plot panel

`eff_linecolor = "slateblue"`: line color in effect plot panel

`eff_linewidth = "3"`: line width in effect plot panel

`eff_xlab = "Genotype"`: x-axis label in effect plot panel

`eff_ylab = "Phenotype"`: y-axis label in effect plot panel

`eff_rotate_ylab = NULL`: indicates whether to rotate the y-axis label 90 degrees, in effect plot panel

`eff_segwidth = NULL`: width of line segments in effect plot panel, in pixels

`eff_axispos = list(xtitle=25, ytitle=30, xlabel=5, ylabel=5)`: position of axis labels in pixels (xtitle, ytitle, xlabel, ylabel) in effect plot panel

`eff_titlepos = 20`: position of title for effect plot panel, in pixels

### `iplotScanone` (`pxgtype="raw"`)

`height = 450`: height of image in pixels

`wleft = 700`: width of left panel in pixels

`wright = 300`: width of right panel in pixels

`margin = list(left=60, top=40, right=40, bottom=40, inner=5)`: margins in pixels (left, top, right, bottom, inner)

`lod_axispos = list(xtitle=25, ytitle=30, xlabel=5, ylabel=5)`: position of axis labels in pixels (xtitle, ytitle, xlabel, ylabel) in LOD curve panel

`lod_titlepos = 20`: position of title for LOD curve panel, in pixels

`chrGap = 8`: gap between chromosomes

`darkrect = "#C8C8C8"`: color of darker background rectangle

`lightrect = "#E6E6E6"`: color of lighter background rectangle

`lod_ylim = NULL`: y-axis limits in LOD curve panel

`lod_nyticks = 5`: number of ticks in y-axis in LOD curve panel

`lod_yticks = NULL`: vector of tick positions for y-axis in LOD curve panel

`lod_linecolor = "darkslateblue"`: line color for LOD curves

`lod_linewidth = 2`: line width for LOD curves

`lod_pointcolor = "#E9CFEC"`: color for points at markers in LOD curve panel

`lod_pointsize = 0`: size of points at markers (default = 0 corresponding to no visible points at markers)

`lod_pointstroke = "black"`: color of outer circle for points at markers in LOD curve panel

`lod_title = ""`: title of LOD curve panel

`lod_xlab = "Chromosome"`: x-axis label for LOD curve panel

`lod_ylab = "LOD score"`: y-axis label for LOD curve panel

`lod_rotate_ylab = NULL`: indicates whether to rotate the y-axis label 90 degrees, in LOD curve panel

`eff_pointcolor = "slateblue"`: point color in phe-by-gen panel

`eff_pointcolorhilit = "Orchid"`: point color, when highlighted, in phe-by-gen panel

`eff_pointstroke = "black"`: color of outer circle for points, in phe-by-gen panel

`eff_pointsize = 3`: point size in phe-by-gen paenl

`eff_ylim = NULL`: y-axis limits in phe-by-gen panel

`eff_nyticks = 5`: number of ticks in y-axis in phe-by-gen panel

`eff_yticks = NULL`: vector of tick positions for y-axis in phe-by-gen panel

`eff_xlab = "Genotype"`: x-axis label in phe-by-gen panel

`eff_ylab = "Phenotype"`: y-axis label in phe-by-gen panel

`eff_rotate_ylab = NULL`: indicates whether to rotate the y-axis label 90 degrees, in phe-by-gen panel

`xjitter = NULL`: amount of horizontal jittering in phe-by-gen panel

`eff_axispos = list(xtitle=25, ytitle=30, xlabel=5, ylabel=5)`: position of axis labels in pixels (xtitle, ytitle, xlabel, ylabel) in LOD curve panel

`eff_titlepos = 20`: position of title for phe-by-gen panel, in pixels

`eff_yNA = list(handle=true, force=false, width=15, gap=10)`: treatment of missing values in phe-by-gen panel (handle=T/F, force=T/F, width, gap)

### `iplotPXG`

`height = 450`: height of chart in pixels

`width = 300`: width of chart in pixels

`title = ""`: title for chart

`margin = list(left=60, top=40, right=40, bottom=40, inner=5)`: margins in pixels (left, top, right, bottom, inner)

`xlab = "Genotype"`: x-axis label

`ylab = "Phenotype"`: y-axis label

`axispos = list(xtitle=25, ytitle=30, xlabel=5, ylabel=5)`: position of axis labels in pixels (xtitle, ytitle, xlabel, ylabel)

`titlepos = 20`: position of chart title in pixels

`xjitter = NULL`: amount of horizontal jittering in pixels

`ylim = NULL`: y-axis limits

`yticks = NULL`: vector of tick positions on y-axis

`nyticks = 5`: no. ticks on y-axis

`rectcolor = "#E6E6E6"`: color of background rectangle

`pointcolor = "slateblue"`: color for points

`pointsize = 3`: size of points in pixels

`pointstroke = "black"`: color of outer circle for points

`yNA = list(handle=true, force=false, width=15, gap=10)`: treatment of missing values (handle=T/F, force=T/F, width, gap)

### `iplotMScanone` (no effects)

`wleft = 650`: width of left panels in pixels

`wright = 350`: width of right panel in pixels

`htop = 350`: height of top panels in pixels

`hbot = 350`: height of bottom panel in pixels

`margin = list(left=60, top=40, right=40, bottom=40, inner=5)`: margins in pixels (left, top, right, bottom, inner)

`axispos = list(xtitle=25, ytitle=30, xlabel=5, ylabel=5)`: position of axis labels in pixels (xtitle, ytitle, xlabel, ylabel)

`titlepos = 20`: position of chart title in pixels

`chrGap = 8`: gap between chromosomes in pixels

`darkrect = "#C8C8C8"`: color of darker background rectangle

`lightrect = "#E6E6E6"`: color of lighter background rectangle

`colors = c("slateblue", "white", "crimson")`: heat map colors

`zlim = NULL`: z-axis limits

`zthresh = NULL`: lower z-axis threshold for display in heat map

`linecolor = "darkslateblue"`: color of lines

`linewidth = 2`: width of lines

### `iplotMScanone` (with effects)

`wleft = 650`: width of left panels in pixels

`wright = 350`: width of right panel in pixels

`htop = 350`: height of top panels in pixels

`hbot = 350`: height of bottom panel in pixels

`margin = list(left=60, top=40, right=40, bottom=40, inner=5)`: margins in pixels (left, top, right, bottom, inner)

`axispos = list(xtitle=25, ytitle=30, xlabel=5, ylabel=5)`: position of axis labels in pixels (xtitle, ytitle, xlabel, ylabel)

`titlepos = 20`: position of chart title in pixels

`chrGap = 8`: gap between chromosomes in pixels

`darkrect = "#C8C8C8"`: color of darker background rectangle

`lightrect = "#E6E6E6"`: color of lighter background rectangle

`colors = c("slateblue", "white", "crimson")`: heat map colors

`zlim = NULL`: z-axis limits

`zthresh = NULL`: lower z-axis threshold for display in heat map

`eff_ylim = NULL`: y-axis limits for effect plot (right panel)

`eff_ylab = ""`: y-axis label for effect plot (right panel)

`linecolor = "darkslateblue"`: line color for LOD curves (lower panel)

`eff_linecolor = NULL`: line color for effect plot (right panel)

`linewidth = 2`: line width for LOD curves (lower panel)

`eff_linewidth = 2`: width of line for effect plot (right panel)


<style type="text/css">
body {
    margin-left: 100px;
}
code {
    padding: 0;
    font-size: 90%;
    color: black;
    background-color: white;
    border: 0px solid white;
}
pre code {
    border: 1px solid #ccc;    
}
</style>
