%\VignetteEngine{knitr::knitr}
%\VignetteIndexEntry{R/qtlcharts User Guide}

# R/qtlcharts User Guide

[R/qtlcharts](http://kbroman.github.io/qtlcharts) is an
[R](http://www.r-project.org) package to create interactive charts for
QTL data, for use with [R/qtl](http://www.rqtl.org). The charts are
saved to a temporary file and opened in a web browser, though you may also
save them to a specific file, to be viewed later.

We are focusing development on the
[Chrome](http://www.google.com/chrome/browser) and
[Safari](https://www.apple.com/safari/) browsers. The graphs may also
be viewed in other browsers, such [Opera](http://www.opera.com/) and
[Firefox](http://www.mozilla.org/en-US/firefox/new/), but it can be
hard to accommodate all possible browser differences.

Information on installing R/qtlcharts is available
[here](http://kbroman.github.io/qtlcharts/pages/installation.html); it
requires installation of the packages [R/qtl](http://www.rqtl.org),
[jsonlite](http://cran.r-project.org/web/packages/jsonlite/), and
[devtools](https://github.com/hadley/devtools).

## Basic use

You first need to load the package.

```{r load_library}
library(qtlcharts)
```

Now let's begin by considering the function `iplotCorr`, which creates
a heatmap of a correlation matrix, linked to scatterplots of the
underlying variables.

We'll first load the `geneExpr` dataset included with the R/qtlcharts package.

```{r geneExpr_data}
data(geneExpr)
```

This is a list with two components. The first, `geneExpr$expr`, is a
`r nrow(geneExpr$expr)` &times; `r ncol(geneExpr$expr)` matrix of gene
expression data. We can use `iplotCorr` with these data, as follows:

```{r iplotCorr_reordered, eval=FALSE}
iplotCorr(geneExpr$expr, reorder=TRUE)
```

This will open ...

![iplotCorr example](Figs/iplotCorr1.png)


Scatterplot colored by genotype, contained in the vector `geneExpr$genotype`.

```{r iplotCorr_reordered_with_genotype, eval=FALSE}
iplotCorr(geneExpr$expr, geneExpr$genotype, reorder=TRUE)
```

![iplotCorr example with scatterplot colored by genotype](Figs/iplotCorr2.png)


## Saving to a file

The interactive graphs produced by R/qtlcharts are, by default, saved
to a temporary file and then opened in the default web browser. If you
want to save a chart to a _particular_ file, just use the `file`
argument, as follows:

```{r iplotCorr_specify_file, eval=FALSE}
iplotCorr(geneExpr$expr, reorder=TRUE, file="~/Desktop/iplotCorr_example.html")
```

The chart will still be opened in the browser; to create the file
without opening it, use the argument `openfile=FALSE`, as follows:

```{r iplotCorr_specify_file_dont_open, eval=FALSE}
iplotCorr(geneExpr$expr, reorder=TRUE, file="~/Desktop/iplotCorr_example.html",
          openfile=FALSE)
```

Charts saved in this way will contain links to resources within the
R/qtlcharts package, and so will be viewable on the same system but
will not be portable to another system that doesn't have R/qtlcharts
installed, or that has the package installed in a different location.

To make the interactive graph portable to other computers (for
example, for sharing with collaborators), use the argument
`onefile=TRUE`, as follows:

```{r iplotCorr_make_portable, eval=FALSE}
iplotCorr(geneExpr$expr, reorder=TRUE, file="~/Desktop/iplotCorr_example.html",
          onefile=TRUE)
```

The file will be somewhat larger, but it will be self-contained and so
may be transferred and viewed on any computer.


## Including a chart in an R Markdown document

The interactive charts may be included within an
[R Markdown](https://www.rstudio.com/ide/docs/r_markdown)-based
report. See the separate [`Rmarkdown` vignette](http://kbroman.github.io/qtlcharts/assets/vignettes/Rmarkdown.html):

```{r Rmarkdown_vignette, eval=FALSE}
vignette("Rmarkdown", "qtlcharts")
```




## Examples of the use of other charts

### `iplotMap`

`iplotMap` creates a (slightly) interactive genetic map. Hover over a
marker position to view the marker name.

```{r iplotMap_example, eval=FALSE}
data(hyper)
map <- pull.map(hyper)

iplotMap(map, title="iplotMap example")
```

### `iplotScanone`

`iplotScanone` creates an interactive chart with LOD
curves linked to estimated QTL effects.


```{r iplotScanone_example, eval=FALSE}
data(hyper)
hyper <- calc.genoprob(hyper, step=1)
out <- scanone(hyper)

iplotScanone(out, hyper, title="iplotScanone example")
```

### `iplotMScanone`

`iplotMScanone` creates a heatmap of LOD curves for a set of genome
scans (for each, for each of a time course of phenotypes), linked to a
plot of the individual genome scans, and also to the QTL effects.

```{r iplotMScanone_example, eval=FALSE}
data(grav)
grav <- calc.genoprob(grav, step=1)
grav <- reduce2grid(grav)
phecol <- seq(1, nphe(grav), by=5)
out <- scanone(grav, phe=phecol)
eff <- estQTLeffects(grav, phe=phecol, what="effects")

iplotMScanone(out, effects=eff,
              title="iplotMScanone example",
              chartOpts=list(eff_ylab="QTL effect"))
```


### `iplotCorr`

`iplotCorr` creates an heat map of a correlation matrix, linked to the
underlying scatterplots.

```{r iplotCorr_example, eval=FALSE}
data(geneExpr)

iplotCorr(geneExpr$expr, geneExpr$genotype,
          title = "iplotCorr example",
          chartOpts=list(cortitle="Correlation matrix",
                         scattitle="Scatterplot"))
```


### `iplotCurves`

`iplotCurves` creates a plot of a set of curves linked to one or two
scatterplots.

```{r iplotCurves_example, eval=FALSE}
data(grav)
times <- attr(grav, "time")
phe <- grav$pheno

iplotCurves(phe, times, phe[,times==2 | times==4], phe[,times==4 | times==6],
            title="iplotCurves example",
            chartOpts=list(curves_xlab="Time (hours)", curves_ylab="Root tip angle (degrees)",
                           scat1_xlab="Angle at 2 hrs", scat1_ylab="Angle at 4 hrs",
                           scat2_xlab="Angle at 4 hrs", scat2_ylab="Angle at 6 hrs"))
```

### `iboxplot`

`iboxplot` creates an interactive graph for a large set of box plots
(rendered as lines connecting the quantiles), linked to underlying
histograms.

```{r iboxplot_example, eval=FALSE}
# simulated data
n.ind <- 500
n.gene <- 10000
expr <- matrix(rnorm(n.ind * n.gene, (1:n.ind)/n.ind*3), ncol=n.gene)
dimnames(expr) <- list(paste0("ind", 1:n.ind),
                       paste0("gene", 1:n.gene))

iboxplot(expr, title="iboxplot example",
         chartOpts=list(xlab="Mice", ylab="Gene expression"))
```

### `iheatmap`

`iheatmap` creates an interactive heatmap of a numeric matrix, with
pixels linked to plots of horizontal and vertical slices.

```{r iheatmap_example, eval=FALSE}
# data to display
n <- 101
x <- y <- seq(-2, 2, len=n)
z <- matrix(ncol=n, nrow=n)
for(i in seq(along=x))
  for(j in seq(along=y))
    z[i,j] <- x[i]*y[j]*exp(-x[i]^2 - y[j]^2)

iheatmap(z, x, y, title = "iheatmap example")
```




## Full list of `chartOpts` options

There are a number of ways in which one may wish to customize the
interactive graphs produced by
[R/qtlcharts](http://kbroman.github.io/qtlcharts), such as different
colors or axis labels. Our main approach to allow such customization
is through a function argument, `chartOpts`. We touched on this
briefly above. The following is a complete listing of the
customization options for each chart.

<<insert_chartOpts_here>>

<style type="text/css">
body {
    margin-left: 100px;
}
code {
    padding: 0;
    font-size: 90%;
    color: black;
    background-color: white;
    border: 0px solid white;
}
pre code {
    border: 1px solid #ccc;    
}
</style>
